### PR TITLE
Add rule newline-per-chained-call

### PR DESCRIPTION
**Pull request by georgekaran**
## Summary

Include a rule to break lines when functions (`CallExpression`) have more than one chained call.

**Why do we need a custom rule?**
The ESLint `newline-per-chained-call` is not so flexible to have some behaviors that we thought is necessary.

Examples:
* When the default value of `ignoreChainWithDepth` is surpassed only the following chained calls would break the line:

  ❌  ESLint newline-per-chained-call
  ```tsx
  // Entry code
  foo().bar().baz().quz()
  
  // Output code
  foo().bar().baz()
      .quz()
  ```

  ✅ Croct newline-per-chained-call
  ```tsx
  foo().bar().baz().quz()
  
  // Output code
  foo()
      .bar()
      .baz()
      .quz()
  ```

* When the chained calls were only `MemberExpression`:

  ❌  ESLint newline-per-chained-call
  ```tsx
  // Input code
  '&:focus-visible': {
      borderColor: ({colors}): string => colors.ui.style.shadow.ring.emphasized,
      boxShadow: ring('emphasized', false, '1px'),
  },
  
  // Output code
  '&:focus-visible': {
      borderColor: ({colors}): string => colors
          .ui
          .style
          .shadow
          .ring
          .emphasized,
      boxShadow: ring('emphasized', false, '1px'),
  },
  ```
  
  ✅ Croct newline-per-chained-call
  ```tsx
  // Input code
  '&:focus-visible': {
      borderColor: ({colors}): string => colors.ui.style.shadow.ring.emphasized,
      boxShadow: ring('emphasized', false, '1px'),
  },
  
  // Output code
  '&:focus-visible': {
      borderColor: ({colors}): string => colors.ui.style.shadow.ring.emphasized,
      boxShadow: ring('emphasized', false, '1px'),
  },
  ```

### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by Fryuni — DISMISSED** [review]

**Comment by Fryuni** [comment]
Which scenario do the existing rule do not support that required us to implement a custom one?

**Comment by georgekaran** [comment]
> Which scenario do the existing rule do not support that required us to implement a custom one?

Sorry my bad, I totally forgot to update the description of the PR.
I've added some examples on why we created a custom rule.

File: src/rules/newline-per-chained-call/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        {code: 'const a = m1.m2(); let b = m1.m2();\nvar c = m1.m2()'},
+        {code: 'const a = m1()\n.m2();'},
+        {code: 'const a = m1();'},
+        {code: 'a().b().c'},
+        {code: 'const foo = a.b.c.e.d'},
+        {code: "a.b.c.e.d = 'foo'"},
+        {code: 'a().b().c()'},
+        {code: 'const a = window\n.location\n.href\n.match(/(^[^#]*)/)[0];'},
+        {code: "const a = window['location']\n.href\n.match(/(^[^#]*)/)[0];"},
+        {code: "const a = window['location'].href.match(/(^[^#]*)/)[0];"},
```
Comment (marcospassos) [inline src/rules/newline-per-chained-call/index.test.ts:27]: Some examples are using `\n` (which I find better than using multiline strings, btw), while other rules are using multiline strings. Can we stick to a single form?

File: src/rules/newline-per-chained-call/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+            ],
+        },
+        {
+            code: '_\n.chain({})\n.map(foo)\n.filter(bar).value();',
+            output: `_
+.chain({})
+.map(foo)
+.filter(bar)
+.value();`,
+            errors: [
```
Comment (marcospassos) [inline src/rules/newline-per-chained-call/index.test.ts:79]: same

File: src/rules/newline-per-chained-call/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+                },
+            ],
+        },
+        {
+            code: 'a.b.c().e().d()',
+            output: `a
+.b
+.c()
+.e()
+.d()`,
```
Comment (marcospassos) [inline src/rules/newline-per-chained-call/index.test.ts:123]: same

File: src/rules/newline-per-chained-call/index.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+            return prefix + lines[0];
+        }
+
+        function isTokenOnSameLine(
+            left: TSESTree.LeftHandSideExpression,
+            right: TSESTree.Expression | TSESTree.PrivateIdentifier,
+        ): boolean {
+            return left.loc.end.line === right.loc.start.line;
+        }
```
Comment (marcospassos) [inline src/rules/newline-per-chained-call/index.ts:58]: I see no reason for this function. It's used only once.

File: src/rules/newline-per-chained-call/index.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            }
+        }
+
+        return {
+            CallExpression: (node: TSESTree.CallExpression): void => {
+                if (node.callee == null || node.callee.type !== 'MemberExpression') {
+                    return;
+                }
+
+                validateCallExpressionIgnoreDepth(node);
```
Comment (marcospassos) [inline src/rules/newline-per-chained-call/index.ts:148]: Better move it to inside the branch. It's more readable since you don't have extra cases.

**Review by marcospassos — CHANGES_REQUESTED** [review]

**Review by marcospassos — CHANGES_REQUESTED** [review]
Please include docs for the new rule.

**Comment by marcospassos** [comment]
If we decide to adopt the `\n`, please fix the other rule's style in another PR as it's unrelated to this.

**Review by Fryuni — APPROVED** [review]

**Pull request merged**